Ветвления If...Else и Loop

avightclav opened the pull request:
В IfElse.kt (lesson2) выполнено три первых задачи, задача triangleKind простого уровня, все задачи среднего уровня.

В Logical.kt (lesson2) выполнены все задачи.

В Loop.kt (lesson3) выполны все тривиальные задачи, четыре первые задачи простого уровня, все задачи среднего уровня, все задачи сложного уровня.

avightclav committed:
...

avightclav committed:
Я сделал большинство задач первого урока.

avightclav committed:
Я сделал большинство задач первого урока.

avightclav committed:
В IfElse.kt (lesson2) выполнено три первых задачи, задача triangleKind простого уровня, все задачи среднего уровня.

В Logical.kt (lesson2) выполнены все задачи.

В Loop.kt (lesson3) выполны все тривиальные задачи, четыре первые задачи простого уровня, все задачи среднего уровня, все задачи сложного уровня.

avightclav committed:
На самом деле предыдущий комит содержал ошибки, теперь они исправлены.

avightclav committed:
Ещё немного исправлений.

belyaev-mikhail commented on src/lesson1/task1/Simple.kt:
```diff
@@ -51,7 +52,7 @@ fun main(args: Array<String>) {
  * Рассчитать время в секундах, прошедшее с начала суток (30035 в данном случае).
  */
-fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
+fun seconds(hours: Int, minutes: Int, seconds: Int): Int = hours * 3600 + minutes * 60 + seconds
```
Длинные тела функции лучше разносить на несколько строк.
Наилучшим решением в данном случае будет поставить перенос строки после знака `=`.

belyaev-mikhail commented on src/lesson1/task1/Simple.kt:
```diff
@@ -60,7 +61,7 @@ fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
  * 1 сажень = 3 аршина = 48 вершков, 1 вершок = 4.445 см.
  */
-fun lengthInMeters(sagenes: Int, arshins: Int, vershoks: Int): Double = TODO()
+fun lengthInMeters(sagenes: Int, arshins: Int, vershoks: Int): Double = ((sagenes * 48 + arshins * 16 + vershoks) * 4.445) / 100
```
Аналогично

belyaev-mikhail commented on src/lesson1/task1/Simple.kt:
```diff
@@ -93,7 +94,7 @@ fun thirdDigit(number: Int): Int = TODO()
  * Определите время поезда в пути в минутах (в данном случае 216).
  */
-fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = TODO()
+fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = (hoursArrive * 60 + minutesArrive) - (hoursDepart * 60 + minutesDepart)
```
Аналогично

belyaev-mikhail commented on src/lesson1/task1/Simple.kt:
```diff
@@ -110,4 +111,4 @@ fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
  *Необходимо вывести число, полученное из заданного перестановкой цифр в обратном порядке (например, 874).
  */
-fun numberRevert(number: Int): Int = TODO()
+fun numberRevert(number: Int): Int = (number % 100 % 10) * 100 + (number % 100 - (number % 100 % 10)) + (number - number % 100) / 100
```
Аналогично

belyaev-mikhail commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,8 +35,21 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+fun ageDescription(age: Int): String {
+    if ((age > 0) && (age < 200)) {
+        return when {
+            age / 10 == 1 || age % 10 == 0 ||  age / 10 == 11 || age % 10 in 5..10 -> "$age лет"
```
Вам не кажется, что этот набор условий можно выразить проще?

belyaev-mikhail commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -44,7 +59,14 @@ fun ageDescription(age: Int): String = TODO()
+    if (S < (t1 * v1) + (t2 * v2)) return (t1 + (S - v1 * t1) / v2)
+    if (S < (t1 * v1) + (t2 * v2) + (t3 * v3)) return t1 + t2 + (S - (v1 * t1 + v2 * t2)) / v3
+
+    return Double.NaN
```
Что вы хотели этим сказать? Подумайте, что такое Double.NaN и зачем вы его возвращаете.

belyaev-mikhail commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -79,7 +112,35 @@ fun rookOrBishopThreatens(kingX: Int, kingY: Int,
+        if (maxSqr < catheusSqr) return 0
+        if (maxSqr == catheusSqr) return 1
+        if (maxSqr > catheusSqr) return 2
+    } else return -1
```
Если ветка `else` возвращает то же самое, что и выражение после неё, зачем её писать?

belyaev-mikhail commented on src/lesson2/task2/Logical.kt:
```diff
@@ -1,31 +1,48 @@
  */
-fun isNumberHappy(number: Int): Boolean = TODO()
+fun isNumberHappy(number: Int): Boolean {
+    val third = number % 1000 % 100 % 10
```
Множество операций взятия остатка при кратных друг другу делителях бессмысленно.
`N % 100 % 10` всегда равно `N % 10` для любых `N`.

belyaev-mikhail replied:
Ваши имена переменных не соответствуют их смыслу. Например, в переменной `third` вы храните _четвёртую_ цифру. Имена переменных должны соответствовать тому, что в них хранится.

belyaev-mikhail commented on src/lesson2/task2/Logical.kt:
```diff
@@ -35,7 +52,7 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
  */
 fun circleInside(x1: Double, y1: Double, r1: Double,
-                 x2: Double, y2: Double, r2: Double): Boolean = TODO()
+                 x2: Double, y2: Double, r2: Double): Boolean = Math.sqrt(sqr(x1 - x2) + sqr(y1 - y2)) + r1 <= r2
```
См. замечание про длину тела функции.

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -1,6 +1,9 @@
+
 package lesson3.task1
 
+import javafx.scene.input.InputMethodTextRun
```
Что это за импорт? Зачем он нужен?
Для удаления неиспользуемых или случайно добавленных импортов в IntelliJ IDEA есть функция `Optimize Imports`

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -34,7 +37,7 @@ fun isPrime(n: Int): Boolean {
 fun isPerfect(n: Int): Boolean {
     var sum = 1
-    for (m in 2..n/2) {
+    for (m in 2..n / 2) {
```
Неконсистентные отступы. Используйте автоформатирование или форматируйте свой код корректно.

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +60,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun digitNumber(n: Int): Int = TODO()
+fun digitNumber(n: Int): Int =
+        if (n / 10 == 0 || n == 0) 1 else digitNumber(n / 10) + digitNumber(n % 10)
```
Какой смысл вызывать `digitNumber(n % 10)`, если он _всегда_ возвращает `1`?

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +60,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  * Ряд Фибоначчи определён следующим образом: fib(1) = 1, fib(2) = 1, fib(n+2) = fib(n) + fib(n+1)
  */
-fun fib(n: Int): Int = TODO()
+fun fib(n: Int): Int = if (n == 1 || n == 2) 1 else fib(n - 1) + fib(n - 2)
```
Можно ли реализовать вычисление числа фибоначчи более эффективно?

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +60,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    while (n % i != 0) {
+        i--
+    }
+
```
Лишних пустых строк следует избегать. Одиночные пустые строки допустимы для отделения друг от друга смысловых частей кода. Несколько пустых строк подряд &mdash; признак плохого стиля.

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -131,7 +160,32 @@ fun cos(x: Double, eps: Double): Double = TODO()
+    var cn = n
+    var rn = 0L
+    while (i != 0) {
+        rn += cn % 10 * Math.pow(10.0, i.toDouble() - 1).toLong()
```
Не следует переводить целые числа в вещественные только для того, чтобы возвести число в степень. Это потенциально может привести к ошибкам округления и другим нежелательным эффектам. Кроме того, это работает крайне медленно. Реализуйте собственную функцию `pow` для целых чисел, это несложно.

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -131,7 +160,32 @@ fun cos(x: Double, eps: Double): Double = TODO()
+    var cn = n
+    var rn = 0L
+    while (i != 0) {
+        rn += cn % 10 * Math.pow(10.0, i.toDouble() - 1).toLong()
```
Аналогично

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +194,20 @@ fun revert(n: Int): Int = TODO()
  */
-fun hasDifferentDigits(n: Int): Boolean = TODO()
+fun hasDifferentDigits(n: Int): Boolean {
+    if (n % 10 != n / 10 % 10 && n / 10 != 0) return true else if (n / 10 == 0) return false else return hasDifferentDigits(n / 10)
```
См. комментарий про длину тела функции.

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +216,17 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+    }
+    val overage = sequence - n
+    val digit = i * i
+    return (digit / Math.pow(10.0, overage.toDouble()) % 10).toInt()
```
См. комментарий про возведение в степень.

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -166,4 +235,34 @@ fun squareSequenceDigit(n: Int): Int = TODO()
+        else {
+            val fixDigit = digitNumber(fib)
+            for (i in 1..fixDigit) {
+                fibList.add(fib / Math.pow(10.0, fixDigit.toDouble()-i).toInt() % 10)
```
См. комментарий про возведение в степень.

belyaev-mikhail commented:
recheck assignment

ice-phoenix commented:
recheck assignment

avightclav commented:
Огромное спасибо за полезные советы, я всё исправлю.

23 сентября 2016 г., 16:59 пользователь Mikhail Belyaev <
notifications@github.com> написал:

> ## _@belyaev-mikhail_ requested changes on this pull request.
> 
> In src/lesson1/task1/Simple.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -51,7 +52,7 @@ fun main(args: Array<String>) {
> > - Пользователь задает время в часах, минутах и секундах, например, 8:20:35.
> > - Рассчитать время в секундах, прошедшее с начала суток (30035 в данном случае).
> >   */
> >   -fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
> >   +fun seconds(hours: Int, minutes: Int, seconds: Int): Int = hours \* 3600 + minutes \* 60 + seconds
> 
> Длинные тела функции лучше разносить на несколько строк.
> Наилучшим решением в данном случае будет поставить перенос строки после
> 
> ## знака =.
> 
> In src/lesson1/task1/Simple.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -60,7 +61,7 @@ fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
> > - Определить длину того же отрезка в метрах (в данном случае 18.98).
> > - 1 сажень = 3 аршина = 48 вершков, 1 вершок = 4.445 см.
> >   */
> >   -fun lengthInMeters(sagenes: Int, arshins: Int, vershoks: Int): Double = TODO()
> >   +fun lengthInMeters(sagenes: Int, arshins: Int, vershoks: Int): Double = ((sagenes \* 48 + arshins \* 16 + vershoks) \* 4.445) / 100
> 
> ## Аналогично
> 
> In src/lesson1/task1/Simple.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -93,7 +94,7 @@ fun thirdDigit(number: Int): Int = TODO()
> > - прибыл на станцию назначения в h2 часов m2 минут того же дня (например в 13:01).
> > - Определите время поезда в пути в минутах (в данном случае 216).
> >   */
> >   -fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = TODO()
> >   +fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = (hoursArrive \* 60 + minutesArrive) - (hoursDepart \* 60 + minutesDepart)
> 
> ## Аналогично
> 
> In src/lesson1/task1/Simple.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -110,4 +111,4 @@ fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
> > - Пользователь задает целое трехзначное число (например, 478).
> >   *Необходимо вывести число, полученное из заданного перестановкой цифр в обратном порядке (например, 874).
> >   */
> >   -fun numberRevert(number: Int): Int = TODO()
> >   +fun numberRevert(number: Int): Int = (number % 100 % 10) \* 100 + (number % 100 - (number % 100 % 10)) + (number - number % 100) / 100
> 
> ## Аналогично
> 
> In src/lesson2/task1/IfElse.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -33,8 +35,21 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
> > - Мой возраст. Для заданного 0 < n < 200, рассматриваемого как возраст человека,
> > - вернуть строку вида: «21 год», «32 года», «12 лет».
> >   */
> >   -fun ageDescription(age: Int): String = TODO()
> >   +fun ageDescription(age: Int): String {
> >   -    if ((age > 0) && (age < 200)) {
> >   -        return when {
> >   -            age / 10 == 1 || age % 10 == 0 ||  age / 10 == 11 || age % 10 in 5..10 -> "$age лет"
> 
> ## Вам не кажется, что этот набор условий можно выразить проще?
> 
> In src/lesson2/task1/IfElse.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -44,7 +59,14 @@ fun ageDescription(age: Int): String = TODO()
> >   */
> >  fun timeForHalfWay(t1: Double, v1: Double,
> >                     t2: Double, v2: Double,
> > -                   t3: Double, v3: Double): Double = TODO()
> > -                   t3: Double, v3: Double): Double {
> > -    val S = ((t1 \* v1) + (t2 \* v2) + (t3 \* v3)) / 2
> > -    if (S < t1 \* v1) return (S / v1)
> > -    if (S < (t1 \* v1) + (t2 \* v2)) return (t1 + (S - v1 \* t1) / v2)
> > -    if (S < (t1 \* v1) + (t2 \* v2) + (t3 \* v3)) return t1 + t2 + (S - (v1 \* t1 + v2 \* t2)) / v3
> >   +
> > -    return Double.NaN
> 
> Что вы хотели этим сказать? Подумайте, что такое Double.NaN и зачем вы его
> 
> ## возвращаете.
> 
> In src/lesson2/task1/IfElse.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > -    if (b == max) {
> > -        min1 = a
> > -        min2 = c
> > -    }
> > -    if (a == max) {
> > -        min1 = b
> > -        min2 = c
> > -    }
> >   +
> > -    if (min1 + min2 > max) {
> > -        val maxSqr = max \* max
> > -        val catheusSqr = min1 \* min1 + min2 \* min2
> > -        if (maxSqr < catheusSqr) return 0
> > -        if (maxSqr == catheusSqr) return 1
> > -        if (maxSqr > catheusSqr) return 2
> > -    } else return -1
> 
> Если ветка else возвращает то же самое, что и выражение после неё, зачем
> 
> ## её писать?
> 
> In src/lesson2/task2/Logical.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> >  /**
> > - Простая
> >   *
> > - Четырехзначное число назовем счастливым, если сумма первых двух ее цифр равна сумме двух последних.
> > - Определить, счастливое ли заданное число, вернуть true, если это так.
> >   */
> >   -fun isNumberHappy(number: Int): Boolean = TODO()
> >   +fun isNumberHappy(number: Int): Boolean {
> >   -    val third = number % 1000 % 100 % 10
> 
> Множество операций взятия остатка при кратных друг другу делителях
> бессмысленно.
> 
> ## N % 100 % 10 всегда равно N % 10 для любых N.
> 
> In src/lesson2/task2/Logical.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> >  /**
> > - Простая
> >   *
> > - Четырехзначное число назовем счастливым, если сумма первых двух ее цифр равна сумме двух последних.
> > - Определить, счастливое ли заданное число, вернуть true, если это так.
> >   */
> >   -fun isNumberHappy(number: Int): Boolean = TODO()
> >   +fun isNumberHappy(number: Int): Boolean {
> >   -    val third = number % 1000 % 100 % 10
> 
> Ваши имена переменных не соответствуют их смыслу. Например, в переменной
> third вы храните _четвёртую_ цифру. Имена переменных должны
> 
> ## соответствовать тому, что в них хранится.
> 
> In src/lesson2/task2/Logical.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -35,7 +52,7 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
> > - Вернуть true, если утверждение верно
> >   */
> >   fun circleInside(x1: Double, y1: Double, r1: Double,
> >   -                 x2: Double, y2: Double, r2: Double): Boolean = TODO()
> >   -                 x2: Double, y2: Double, r2: Double): Boolean = Math.sqrt(sqr(x1 - x2) + sqr(y1 - y2)) + r1 <= r2
> 
> ## См. замечание про длину тела функции.
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> >  package lesson3.task1
> 
> +import javafx.scene.input.InputMethodTextRun
> 
> Что это за импорт? Зачем он нужен?
> Для удаления неиспользуемых или случайно добавленных импортов в IntelliJ
> 
> ## IDEA есть функция Optimize Imports
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -34,7 +37,7 @@ fun isPrime(n: Int): Boolean {
> >   */
> >  fun isPerfect(n: Int): Boolean {
> >      var sum = 1
> > -    for (m in 2..n/2) {
> > -    for (m in 2..n / 2) {
> 
> Неконсистентные отступы. Используйте автоформатирование или форматируйте
> 
> ## свой код корректно.
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -57,37 +60,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
> > - Найти количество цифр в заданном числе n.
> > - Например, число 1 содержит 1 цифру, 456 -- 3 цифры, 65536 -- 5 цифр.
> >   */
> >   -fun digitNumber(n: Int): Int = TODO()
> >   +fun digitNumber(n: Int): Int =
> >   -        if (n / 10 == 0 || n == 0) 1 else digitNumber(n / 10) + digitNumber(n % 10)
> 
> ## Какой смысл вызывать digitNumber(n % 10), если он _всегда_ возвращает 1?
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> >  /**
> > - Простая
> >   *
> > - Найти число Фибоначчи из ряда 1, 1, 2, 3, 5, 8, 13, 21, ... с номером n.
> > - Ряд Фибоначчи определён следующим образом: fib(1) = 1, fib(2) = 1, fib(n+2) = fib(n) + fib(n+1)
> >   */
> >   -fun fib(n: Int): Int = TODO()
> >   +fun fib(n: Int): Int = if (n == 1 || n == 2) 1 else fib(n - 1) + fib(n - 2)
> 
> ## Можно ли реализовать вычисление числа фибоначчи более эффективно?
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> >  /**
> > - Простая
> >   *
> > - Для заданного числа n > 1 найти максимальный делитель, меньший n
> >   */
> >   -fun maxDivisor(n: Int): Int = TODO()
> >   +fun maxDivisor(n: Int): Int {
> >   -    var i = n / 2
> >   -    while (n % i != 0) {
> >   -        i--
> >   -    }
> >     +
> 
> Лишних пустых строк следует избегать. Одиночные пустые строки допустимы
> для отделения друг от друга смысловых частей кода. Несколько пустых строк
> 
> ## подряд — признак плохого стиля.
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -131,7 +160,32 @@ fun cos(x: Double, eps: Double): Double = TODO()
> > - Поменять порядок цифр заданного числа n на обратный: 13478 -> 87431.
> > - Не использовать строки при решении задачи.
> >   */
> >   -fun revert(n: Int): Int = TODO()
> >   +fun revert(n: Int): Int {
> >   -    var i = digitNumber(n)
> >   -    var cn = n
> >   -    var rn = 0L
> >   -    while (i != 0) {
> >   -        rn += cn % 10 \* Math.pow(10.0, i.toDouble() - 1).toLong()
> 
> Не следует переводить целые числа в вещественные только для того, чтобы
> возвести число в степень. Это потенциально может привести к ошибкам
> округления и другим нежелательным эффектам. Кроме того, это работает крайне
> медленно. Реализуйте собственную функцию pow для целых чисел, это
> 
> ## несложно.
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > -    var rn = 0L
> > -    while (i != 0) {
> > -        rn += cn % 10 \* Math.pow(10.0, i.toDouble() - 1).toLong()
> > -        cn /= 10
> > -        i--
> > -    }
> >   +
> > -    return rn.toInt()
> >   +}
> >   +
> >   +fun revertLong(n: Int): Long {
> > -    var i = digitNumber(n)
> > -    var cn = n
> > -    var rn = 0L
> > -    while (i != 0) {
> > -        rn += cn % 10 \* Math.pow(10.0, i.toDouble() - 1).toLong()
> 
> ## Аналогично
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> >  /**
> > - Средняя
> >   *
> > - Для заданного числа n определить, содержит ли оно различающиеся цифры.
> > - Например, 54 и 323 состоят из разных цифр, а 111 и 0 из одинаковых.
> >   */
> >   -fun hasDifferentDigits(n: Int): Boolean = TODO()
> >   +fun hasDifferentDigits(n: Int): Boolean {
> >   -    if (n % 10 != n / 10 % 10 && n / 10 != 0) return true else if (n / 10 == 0) return false else return hasDifferentDigits(n / 10)
> 
> ## См. комментарий про длину тела функции.
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > @@ -157,7 +216,17 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
> > - 149162536496481100121144...
> > - Например, 2-я цифра равна 4, 7-я 5, 12-я 6.
> >   */
> >   -fun squareSequenceDigit(n: Int): Int = TODO()
> >   +fun squareSequenceDigit(n: Int): Int {
> >   -    var sequence = 0
> >   -    var i = 0
> >   -    while (n > sequence) {
> >   -        i++
> >   -        sequence += digitNumber(i \* i)
> >   -    }
> >   -    val overage = sequence - n
> >   -    val digit = i \* i
> >   -    return (digit / Math.pow(10.0, overage.toDouble()) % 10).toInt()
> 
> ## См. комментарий про возведение в степень.
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885
> :
> 
> > -  val overage = sequence - n;
> > -  val digit = fib(i)
> > -  return (digit / Math.pow(10.0, overage.toDouble()) % 10).toInt()
> >   +*/
> > -    val fibList = mutableListOf<Int>()
> > -    var fib = 1
> > -    var fib1 = 0
> > -    var fib2 = 1
> >   +
> >   +
> > -    while (n > fibList.size) {
> > -        if (fib < 10) fibList.add(fib)
> > -        else {
> > -            val fixDigit = digitNumber(fib)
> > -            for (i in 1..fixDigit) {
> > -                fibList.add(fib / Math.pow(10.0, fixDigit.toDouble()-i).toInt() % 10)
> 
> См. комментарий про возведение в степень.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1329885,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AQS_FLokcbE-S4cTYLVaZZLj9Diqkun4ks5qs9s5gaJpZM4KD7sS
> .

avightclav committed:
В IfElse.kt (lesson2) выполнено три первых задачи, задача triangleKind простого уровня, все задачи среднего уровня.

В Logical.kt (lesson2) выполнены все задачи.

В Loop.kt (lesson3) выполны все тривиальные задачи, четыре первые задачи простого уровня, все задачи среднего уровня, все задачи сложного уровня.

avightclav committed:
Исправил, с учётом советов Михаила Беляева.

avightclav committed:
Исправил isNumberHappy и GetAge

belyaev-mikhail commented on src/lesson1/task1/Simple.kt:
```diff
@@ -110,4 +114,5 @@ fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
  */
-fun numberRevert(number: Int): Int = TODO()
+fun numberRevert(number: Int): Int =
+        (number % 100 % 10) * 100 + (number % 100 - (number % 100 % 10)) + (number - number % 100) / 100
```
По-прежнему, `number % 100 % 10` это то же самое, что и просто `number % 10`. Не вполне понятно, зачем вы делаете эту операцию именно так.

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,37 +59,73 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun lcm(m: Int, n: Int): Int {
+    val gcd = gcd(m, n)
+    return m * n / gcd
+
```
Лишние пустые строки

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +203,33 @@ fun revert(n: Int): Int = TODO()
+    else return hasDifferentDigits(n / 10)
+}
+
+fun pow (n: Int, p: Int): Int {
```
Функция написана правильно, но можно лучше.
Исправлять это не требуется для засчитывания задач, но подумайте на досуге, как можно использовать тот факт, что `pow(X, 2*N) == pow(X * X, N) == pow(X, N) * pow(X, N)`, чтобы сделать эту функцию значительно быстрее для чётных степеней путём уменьшения числа умножений. Потом подумайте, как расширить этот же подход на нечётные степени.

avightclav commented:
Спасибо огромное за замечания! Всё что касается первого урока -- по
какой-то причине я совсем забыл как работает деление с остатком (сложно
представить такое, конечно, но со мной случается). Я собирался всё
доделать/переделать, когда все уроки изучу. Кстати, как Вы считаете, для
улучшения умений, лучше ко всему прикоснуться(как я это делаю сейчас), а
потом разбираться глубже или же во всех уроках выполнять все задания. Или
же это выбирать индивидуально, в зависимости от собственных ощущений?

24 сент. 2016 г. 0:05 пользователь "Mikhail Belyaev" <
notifications@github.com> написал:

> ## _@belyaev-mikhail_ requested changes on this pull request.
> 
> In src/lesson1/task1/Simple.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1404428
> :
> 
> > @@ -110,4 +114,5 @@ fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
> > - Пользователь задает целое трехзначное число (например, 478).
> >   *Необходимо вывести число, полученное из заданного перестановкой цифр в обратном порядке (например, 874).
> >   */
> >   -fun numberRevert(number: Int): Int = TODO()
> >   +fun numberRevert(number: Int): Int =
> >   -        (number % 100 % 10) \* 100 + (number % 100 - (number % 100 % 10)) + (number - number % 100) / 100
> 
> По-прежнему, number % 100 % 10 это то же самое, что и просто number % 10.
> 
> ## Не вполне понятно, зачем вы делаете эту операцию именно так.
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1404428
> :
> 
> >  /**
> > - Простая
> >   *
> > - Для заданных чисел m и n найти наименьшее общее кратное, то есть,
> > - минимальное число k, которое делится и на m и на n без остатка
> >   */
> >   -fun lcm(m: Int, n: Int): Int = TODO()
> >   +fun lcm(m: Int, n: Int): Int {
> >   -    val gcd = gcd(m, n)
> >   -    return m \* n / gcd
> >     +
> 
> ## Лишние пустые строки
> 
> In src/lesson3/task1/Loop.kt
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1404428
> :
> 
> >  /**
> > - Средняя
> >   *
> > - Для заданного числа n определить, содержит ли оно различающиеся цифры.
> > - Например, 54 и 323 состоят из разных цифр, а 111 и 0 из одинаковых.
> >   */
> >   -fun hasDifferentDigits(n: Int): Boolean = TODO()
> >   +fun hasDifferentDigits(n: Int): Boolean {
> >   -    if (n % 10 != n / 10 % 10 && n / 10 != 0) return true
> >   -    else if (n / 10 == 0) return false
> >   -    else return hasDifferentDigits(n / 10)
> >     +}
> >     +
> >     +fun pow (n: Int, p: Int): Int {
> 
> Функция написана правильно, но можно лучше.
> Исправлять это не требуется для засчитывания задач, но подумайте на
> досуге, как можно использовать тот факт, что pow(X, 2*N) == pow(X \* X, N)
> == pow(X, N) \* pow(X, N), чтобы сделать эту функцию значительно быстрее
> для чётных степеней путём уменьшения числа умножений. Потом подумайте, как
> расширить этот же подход на нечётные степени.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-1404428,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AQS_FGMh7lyKOP-pA3HMZCziVkpQC0aZks5qtD81gaJpZM4KD7sS
> .

belyaev-mikhail commented:
Это сложный вопрос. Если вы хотите максимально быстро достичь результата, то лучше всего набить как можно больше шишек, а для этого следует выполнять все задания вообще, и когда что-то работает неправильно, пытаться детально разобраться, почему так. Это как с ездой на велосипеде или плаванием &mdash; чем больше делаешь, тем лучше получается.
А вот в каком порядке это делать &mdash; это уже индивидуально.

С точки же зрения технической организации этого семестра, лучше, если вы будете сначала исправлять все имеющиеся замечания, а потом делать делать следующие задания новыми реквестами.

avightclav committed:
Исправил в "Lesson1" и "Lesson3" все недочёты, на которые указал Михаил Беляев.

В "Lesson4" сделал все задачи до factorize(включительно)

avightclav committed:
Поправил недочёты в задах четвертого урока.

belyaev-mikhail commented:
>    [Normal] lesson4.task1/accumulate
>        org.opentest4j.AssertionFailedError : You should mutate an input list, not create a copy

Это сообщение говорит о том, что в задании 4 нужно возвращать тот же список, что вам передали на вход, а не его копию. Модифицировать, соответственно, тоже нужно именно его.

avightclav commented:
Я это уже понял. Только зачем???

belyaev-mikhail commented:
Затем, что именно так сформулировано задание. Вы же не задаётесь вопросом "зачем в задании по подсчёту среднего считать среднее, если можно посчитать сумму?"

belyaev-mikhail commented:
recheck all

avightclav committed:
Исправил accumulate и factorize.

avightclav committed:
Исправил factorize.

avightclav commented:
Я сделял программу с записью в файл!!!!

avightclav commented:
Так, а на пятерку нужно ещё сложную сделать из коллекций, если я уже из
других уроков делал сложные?

mglukhikh commented:
"5" за сентябрь ставится, если у вас решены какие-то задачи из минимум трёх уроков, и хотя бы одна из ВСЕХ решённых задач отмечена как "Сложная".

mglukhikh commented:
Насколько я вижу, вы этим требованиям удовлетворяете :)

belyaev-mikhail commented:
Вы сделали программу с записью в файл, но я, хоть убейте, не понимаю, зачем.
Ваше решение больше затратит на чтение-запись чисел из файла, чем тупой перебор затратил бы на пропуск "непростых" чисел.

belyaev-mikhail commented on src/lesson4/task1/List.kt:
```diff
@@ -157,7 +203,66 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+        }
+    }
+    else factorized.add(n)
+    print(factorized)
```
Функций отладочной печати не должно быть в решении, которое вы отправляете на проверку. Чистите от них код перед отправкой.

belyaev-mikhail commented on src/lesson4/task1/List.kt:
```diff
@@ -128,7 +154,13 @@ fun center(list: MutableList<Double>): MutableList<Double> = TODO()
  */
-fun times(a: List<Double>, b: List<Double>): Double = TODO()
+fun times(a: List<Double>, b: List<Double>): Double {
+    var C = 0.0
```
Стиль кода должен быть консистентным. Если вы хотите называть переменные и параметры с большой буквы &mdash; пожалуйста, но делайте это тогда _везде_.

belyaev-mikhail commented:
recheck all

avightclav commented:
Я просто подумал, что было бы очень круто, чтобы она каждый раз получая
бОльшие значения набирала бы свой несгораемый запас, который бы не
приходилось вычислять заного. А то, что это дольше, я, конечно же, не знал.

Да и это отличная возможность была для меня узнать новое.

29 сент. 2016 г. 21:18 пользователь "Mikhail Belyaev" <
notifications@github.com> написал:

Вы сделали программу с записью в файл, но я, хоть убейте, не понимаю, зачем.
Ваше решение больше затратит на чтение-запись чисел из файла, чем тупой
перебор затратил бы на пропуск "непростых" чисел.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#issuecomment-250548512,
or mute the thread
https://github.com/notifications/unsubscribe-auth/AQS_FJXgmgzG6Imkafth7rRpRGERp6Iyks5qvADmgaJpZM4KD7sS
.

avightclav commented:
Опусы с переменной и отладочным выводом обязательно поправлю!

Ещё я был бы несказанно рад, если бы Вы мне подсказали где можно прочитать
про таймеры/"измерители нагрузок на компьютер", чтобы можно было сравнивать
свои подходы! Спасибо!

29 сент. 2016 г. 21:22 пользователь "Mikhail Belyaev" <
notifications@github.com> написал:

## _@belyaev-mikhail_ requested changes on this pull request.

In src/lesson4/task1/List.kt
https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-2199538
:

> -    var nCopy = n
> -    var i = 0
> -    val factorized = mutableListOf<Int>()
> -    val primes = primeNumbersReader().filter { n % it == 0 }
> -    if (primes.size != 0) {
> -        while (nCopy != 1) {
> -            if (nCopy % primes[i] == 0) {
> -                nCopy /= primes[i]
> -                factorized.add(primes[i])
> -            } else {
> -                i++
> -            }
> -        }
> -    }
> -    else factorized.add(n)
> -    print(factorized)

Функций отладочной печати не должно быть в решении, которое вы отправляете

## на проверку. Чистите от них код перед отправкой.

In src/lesson4/task1/List.kt
https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-2199538
:

> @@ -128,7 +154,13 @@ fun center(list: MutableList<Double>): MutableList<Double> = TODO()
> - представленные в виде списков a и b. Скалярное произведение
>   считать по формуле:
> - C = a1b1 + a2b2 + ... + aNbN. Произведение пустых векторов считать
>   равным 0.0.
>   */
>   -fun times(a: List<Double>, b: List<Double>): Double = TODO()
>   +fun times(a: List<Double>, b: List<Double>): Double {
>   -    var C = 0.0

Стиль кода должен быть консистентным. Если вы хотите называть переменные и
параметры с большой буквы — пожалуйста, но делайте это тогда _везде_.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#pullrequestreview-2199538,
or mute the thread
https://github.com/notifications/unsubscribe-auth/AQS_FBwQf91MPfieH0QCVsmg28qWqwvNks5qvAIAgaJpZM4KD7sS
.

belyaev-mikhail commented:
Для простых случаев (только померять время) вы можете замерять время работы прямо в коде, как предлагают, например, [вот тут](http://stackoverflow.com/questions/3382954/measure-execution-time-for-a-java-method). В стандартной библиотеке Kotlin есть функция `measureTimeMillis`, которая позволяет легко померять время работы произвольного кода:

```
val time = measureTimeMillis{
    runLongFunction()
    runSomethingElse()
}
println("finished in $time milliseconds")
```

Оценить потребление ресурсов для долго работающей программы (например, сервера), вы можете с помощью программы JVisualVM, которая поставлятся вместе с Java и имеет собственный графический интерфейс.

Правильный и профессиональный способ написания бенчмарков для Java описан [тут](http://openjdk.java.net/projects/code-tools/jmh/) и в руководствах Евгения Шипилёва по описанному средству.

Все эти средства предназначены для Java, но подходят и для программ на Kotlin, потому что механизмы работы у них одинаковые. Специализированных средств для Kotlin, насколько мне известно, пока нет.

avightclav commented:
Ещё раз огромное спасибо!!!!

29 сент. 2016 г. 23:05 пользователь "Mikhail Belyaev" <
notifications@github.com> написал:

> Для простых случаев (только померять время) вы можете замерять время
> работы прямо в коде, как предлагают, например, вот тут
> http://stackoverflow.com/questions/3382954/measure-execution-time-for-a-java-method.
> В стандартной библиотеке Kotlin есть функция measureTimeMillis, которая
> позволяет легко померять время работы произвольного кода:
> 
> val time = measureTimeMillis{
>     runLongFunction()
>     runSomethingElse()
> }
> println("finished in $time milliseconds")
> 
> Оценить потребление ресурсов для долго работающей программы (например,
> сервера), вы можете с помощью программы JVisualVM, которая поставлятся
> вместе с Java и имеет собственный графический интерфейс.
> 
> Правильный и профессиональный способ написания бенчмарков для Java описан
> тут http://openjdk.java.net/projects/code-tools/jmh/ и в руководствах
> Евгения Шипилёва по описанному средству.
> 
> Все эти средства предназначены для Java, но подходят и для программ на
> Kotlin, потому что механизмы работы у них одинаковые. Специализированных
> средств для Kotlin, насколько мне известно, пока нет.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/Kotlin-Polytech/KotlinAsFirst2016/pull/104#issuecomment-250575888,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AQS_FHarbWuyBlKosA0E4W4u7qTbc29Fks5qvBn6gaJpZM4KD7sS
> .

belyaev-mikhail commented:
На будущее: такие вопросы лучше задавать преподавателям на занятиях или по почте. Секция комментариев к pull request'ам для этого подходит плохо.

avightclav commented:
Иногда добиться ответа от Марата Халимовича очень сложно. Можно мне в таком
случае общаться к вам в 301 кабинет?

avightclav commented:
Или узнать вашу почту!

belyaev-mikhail commented:
Вообще-то, я тоже веду занятия у вашей группы, если я вас ни с кем не путаю.
Я так понимаю, деление на подгруппы у вас весьма условно, поэтому вы можете задавать свои вопросы на паре мне. В прошлый понедельник я сидел с половиной вашей группы в ауд. 309.
Моя почта есть на сайте нашей кафедры.

avightclav committed:
Сделал пять первых задач пятого урока. (ПЯЯЯЯТЬ! ПЯЯЯТЬ!)

avightclav committed:
Исправил ПАТЕРНЬ в задании bestHighJump

avightclav committed:
Сделал программу plusMinus.
ПРОСТИТЕ МЕНЯ ЗА ЭТУ ТУПУЮ ПРОГРАММУ, НО ОНА РАБОТАЕТ (скорее всего).

avightclav commented:
ЧТО ДЕЛАТЬ??? МОЙ ПРЕПОДАВАТЕЛЬ МАРАТ ХАЛИМОВИЧ (МЕЖДУ ПРОЧИМ ОЧЕНЬ АВТОРИТЕТНЫЙ ЧЕЛОВЕК!!!!) СКАЗАЛ МНЕ, ЧТО ПРЫЖКИ ПОВТОРЯТЬСЯ НЕ МОГУТ. А ТУТ ОНИ ПОВТОРЯЮТСЯЯ!!!! ЧТО ДЕЛАТЬЬ??

belyaev-mikhail commented:
Успокоиться, выпить успокаивающего (или водки), собраться с мыслями и исправить соответствующую функцию.

belyaev-mikhail commented:
Я, правда, не понял, что именно здесь повторяется.

avightclav committed:
Оптимизировал программу plusMinus.

avightclav commented:
Прыжок 147483647 в проверке участвует два раза.

avightclav commented:
Не в проверке, а в исходных данных.

avightclav committed:
Оптимизировал программу plusMinus.

belyaev-mikhail commented:
Я не знаю, чего вам сказал авторитетный человек Марат Халимович, но результаты прыжков абсолютно точно могут повторяться. Кроме того, ваша программа неправильно работает, судя по всему, не поэтому.

avightclav commented:
А со StackOverflow что делать? Как узнать в результате какой строки он хотя бы появляется?

belyaev-mikhail commented:
Вообще, StackOverflow это признак зашедшей &laquo;не туда&raquo; рекурсии. Но ваша функция рекурсии не содержит, из чего я делаю вывод, что вы как-то **очень** сильно сделали худо движку регулярных выражений. Следует проверить правильность ваших регексов, вы как-то, судя по всему, на них сбацали бесконечный (или, по крайней мере, очень глубокий) цикл.

belyaev-mikhail commented:
Конечно, очень неудобно, что бот не вывел вам, на каком именно входе это произошло, но такова природа ошибок типа `StackOverflow` &mdash; зачастую после них вообще ничего невозможно ни увидеть, ни понять.

avightclav commented:
Конечно по этому! Разве могу я теперь в карту, в качестве ключей записывать длину прыжков?

belyaev-mikhail commented:
recheck assignment

avightclav committed:
Исправил flattenPhoneNumber.

belyaev-mikhail commented:
Ну ок, поэтому.
Но результаты прыжков всё равно могут повторяться, так что ваш подход с картой некорректен.
Нужно либо сначала фильтровать, а потом в неё складывать (тогда неважно, что они повторяются, перезапишется и чёрт с ним), либо делать вообще всё как-то по-другому.

belyaev-mikhail commented:
Алсо, проблема со StackOverflow была вызвана очень длинной корректной строкой на входе.
Больше такие строки он генерировать не должен.
Мораль: не играйте с регексами.

belyaev-mikhail commented:
recheck assignment

belyaev-mikhail commented:
recheck assignment

avightclav committed:
Испрвил bestHighJump

belyaev-mikhail commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +79,18 @@ fun timeForHalfWay(t1: Double, v1: Double,
+    val YY1 = kingY == rookY1
+    val YY2 = kingY == rookY2
+    return when {
+        (XX1 && XX2) || (YY1 && YY2) || (XX1 && YY1) || (XX1 && YY2) || (XX2 && YY1) || (XX2 && YY2) == true -> 3
```
`X == true` не имеет смысла, это то же самое, что и просто `X`

Кроме того, вы неправильно понимаете собой же написанное.
`X || Y == true` это не `(X || Y) == true`, а `X || (Y == true)`, что ещё более бессмысленно.

belyaev-mikhail commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +65,15 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
+        c -> min2 = Math.min(a, b)
+    }
+
+    if (min1 <= r && min2 <= s || min2 <= r && min1 <= s) return true else return false
```
`if(X) return true else return false` это то же самое, что и просто `return X`

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +200,40 @@ fun revert(n: Int): Int = TODO()
+fun hasDifferentDigits(n: Int): Boolean {
+    if (n % 10 != n / 10 % 10 && n / 10 != 0) return true
+    else if (n / 10 == 0) return false
+    else return hasDifferentDigits(n / 10)
```
`if`-`else if`-`else if` лучше заменить на `when`

belyaev-mikhail commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,15 +200,40 @@ fun revert(n: Int): Int = TODO()
+    for (i in 2..p) {
+        np *= n
+        if (p % 2 == 0 && p / 2 == i)
+            return np*np
```
Пробелы

belyaev-mikhail commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +63,28 @@ fun main(args: Array<String>) {
-fun dateStrToDigit(str: String): String = TODO()
+fun dateStrToDigit(str: String): String {
+    val partsOfDate = str.split(" ")
+    val months = listOf("Января", "Февраля", "Марта", "Апреля", "Мая", "Июня", "Июля", "Августа", "Сентября", "Октября", "Ноября", "Декабря")
```
Длинная строка, разбейте

belyaev-mikhail commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +124,16 @@ fun dateDigitToStr(digital: String): String = TODO()
+
+fun flattenPhoneNumber(phone: String): String {
+    try {
+        if (!(phone.matches(Regex("[0-9()+-[\\s]]*[0-9]+")))) throw IllegalArgumentException()
```
Вы сами-то понимаете, что эти регексы делают? :-)

avightclav replied:
Понимаю. До этого программа не прошла проверку со строкой ")", т. к. ".martches" пропустил её, поэтому я просто добавил условие того, чтобы в строке просто была цифра хотя бы!

belyaev-mikhail commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,18 +168,52 @@ fun bestLongJump(jumps: String): Int = TODO()
+        val jumpsList = jumps.split(" ")
+        val jumpsInt = jumpsList.filterIndexed { i, s -> i % 2 == 0 }.map(String::toInt)
+        val jumpsSymb = jumpsList.filterIndexed { i, s -> i % 2 != 0 }
+        return jumpsInt.zip(jumpsSymb).filter { it.second.contains(Regex(".*[+].*")) }.toMap().toSortedMap(Comparator { o1, o2 -> o2 - o1 }).filter { it.value.matches(Regex(".*[+]+.*")) }.keys.firstOrNull() ?: -1
```
1. `bla.toMap().toSortedMap()` &mdash; это зачем? `toSortedMap()` тоже отсеивает дубликаты и является `Map`.
2. Зачем вы передаёте свой компаратор? Чем вас стандартный не устраивает, он ведь точно так же работает?
3. Второй фильтр не нужен после того, как вы уже сделали первый.

avightclav replied:
1. Сделать сразу .toSortedMap() нельзя, я проверил несколько раз. Проверьте!
2. Стандартный компаратор располагает значения от меньшего к большему, этот -- не стандартный, он располагает значения от большего к меньшему.
3. Второй фильтр недоглядел, конечно же.

belyaev-mikhail commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,18 +168,52 @@ fun bestLongJump(jumps: String): Int = TODO()
+    if (!(expression.matches(pattern))) throw IllegalArgumentException()
+    val digitsAndSymbols = expression.split(" ")
+    val digits = digitsAndSymbols.filterIndexed { i, s -> i % 2 == 0  }
+    val symbols = listOf<String>("+").plus(digitsAndSymbols.filterIndexed { i, s -> i != 0; i % 2 != 0 })
```
`{ i, s -> i != 0; i % 2 != 0 }` это чего такое? Вы, наверное, имели в виду `{ i, s -> i != 0 && i % 2 != 0 }`?

avightclav replied:
Совершенно верно! Странно, что это тем не менее работало...

belyaev-mikhail replied:
Из этого я только могу сделать вывод, что первое выражение не нужно.

avightclav committed:
Исправил программы, с учётом пожеланий выше.

avightclav committed:
Исправил недоразумение в программе plusMinus.

avightclav commented:
Срочный вопрос. Почему нельзя отдельную часть строки превратить в целое число? Например, существует переменная S значение которой "30", при вызове S[0].toInt выходят какие-то странные рандомные числа. Как с этим бороться? Неужели каждый раз заводить новые переменнные? V = S[0]; V = V.toInt()?

avightclav commented:
Всё разобрался почему! Сделал отдельную функцию, которая делает .toStr().toInt()

avightclav committed:
Сделал задачу уровня "Impossible" в четвертом уроке.

avightclav committed:
Сделал задачу уровня "Impossible" в четвертом уроке.

avightclav committed:
Исправил задачу уровня "Impossible" в четвертом уроке.

avightclav committed:
Исправил задачу уровня "Impossible" в четвертом уроке.

avightclav committed:
Сделал задачу firstDuplicateIndex в пятом уроке.

belyaev-mikhail commented:
Я закрою этот PR, чтобы он не становился вечным. Но учтите указанные замечания на будущее.
